// @kadena/cryptography-utils/lib/tests/hash.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const hash_1 = require("../hash");
const pactTestCommand = {
    networkId: null,
    payload: {
        exec: {
            data: {
                'accounts-admin-keyset': [
                    'ba54b224d1924dd98403f5c751abdd10de6cd81b0121800bf7bdbdcfaec7388d',
                ],
            },
            code: '(define-keyset \'k (read-keyset "accounts-admin-keyset"))\n(module system \'k\n  (defun get-system-time ()\n    (time "2017-10-31T12:00:00Z")))\n(get-system-time)',
        },
    },
    signers: [
        {
            pubKey: 'ba54b224d1924dd98403f5c751abdd10de6cd81b0121800bf7bdbdcfaec7388d',
        },
    ],
    meta: {
        creationTime: 0,
        ttl: 0,
        gasLimit: 0,
        chainId: '',
        gasPrice: 0,
        sender: '',
    },
    nonce: JSON.stringify('step01'),
};
(0, vitest_1.test)('should perform blake2b256 hashing on a string and output hash encoded as unescaped base64url', () => {
    const str = JSON.stringify(pactTestCommand);
    const actualHash = (0, hash_1.hash)(str);
    (0, vitest_1.expect)(actualHash).toEqual('uolsidh4DWN-D44FoElnosL8e5-cGCGn_0l2Nct5mq8');
});
